File: app/src/main/java/com/android/ui/kent/demo/network/retrofit/RetrofitActivity.java
package com.android.ui.kent.demo.network.retrofit;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.widget.Toast;

import com.android.ui.kent.R;
import com.android.ui.kent.demo.BaseActivity;
import com.android.ui.kent.demo.network.retrofit.vo.Repo;
import com.android.ui.kent.demo.network.retrofit.vo.User;
import com.google.gson.Gson;

import java.util.List;

import butterknife.ButterKnife;
import butterknife.OnClick;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by Kent on 2016/11/18.
 */

public class RetrofitActivity extends BaseActivity implements Callback<List<Repo>> {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_retrofit);

        ButterKnife.bind(this);
        initToolbar();
    }

    private void initToolbar() {
        this.setupToolbar();
        this.setToolbarTitle("Retrofit");
        this.enableBackButton();
    }


    @OnClick(R.id.btn_repo)
    void onBtnRepoClick(){

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("https://api.github.com/")
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        GitHubService service = retrofit.create(GitHubService.class);
        Call<List<Repo>> repos = service.listRepos("kentsong");
        repos.enqueue(this);
//        repos.execute();

    }

    @OnClick(R.id.btn_user)
    void onBtnUserClick(){

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("https://api.github.com/")
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        GitHubService service = retrofit.create(GitHubService.class);
        Call<List<User>> repos = service.listUser();
        repos.enqueue(new Callback<List<User>>() {
            @Override
            public void onResponse(Call<List<User>> call, Response<List<User>> response) {
                Toast.makeText(RetrofitActivity.this, "onResponse json = " + new Gson().toJson(response), Toast.LENGTH_LONG).show();
            }

            @Override
            public void onFailure(Call<List<User>> call, Throwable t) {
                Toast.makeText(RetrofitActivity.this, "onFailure", Toast.LENGTH_SHORT).show();
            }
        });

    }

    @Override
    public void onResponse(Call<List<Repo>> call, Response<List<Repo>> response) {
        Toast.makeText(this, "onResponse json = " + new Gson().toJson(response), Toast.LENGTH_LONG).show();
        List<Repo> list = response.body();
    }


    @Override
    public void onFailure(Call<List<Repo>> call, Throwable t) {
        Toast.makeText(this, "onFailure", Toast.LENGTH_SHORT).show();
    }

    public static void launch(Activity activity) {
        Intent intent = new Intent(activity, RetrofitActivity.class);
        activity.startActivity(intent);
    }
}

File: slimsample/src/main/java/com/kent/android/slim/sample/retorfit/RetrofitActivity.kt
package com.kent.android.slim.sample.retorfit

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.os.Parcelable
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import com.google.gson.Gson
import com.kent.android.slim.sample.R
import com.kent.android.slim.sample.rejoin.*
import kotlinx.android.parcel.Parcelize
import kotlinx.android.synthetic.main.activity_eventbus.*
import kotlinx.android.synthetic.main.activity_eventbus.view.*
import okhttp3.OkHttpClient
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.*
import java.security.SecureRandom
import java.security.cert.CertificateException
import java.security.cert.X509Certificate
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager
import javax.net.ssl.X509TrustManager

/**
 */
class RetrofitActivity : AppCompatActivity() {
    private val TAG = RetrofitActivity::class.java.simpleName
    private var host: Host? = null

    private val client: OkHttpClient by lazy {
        getTrustAllBuilder().build()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_eventbus)

        val retrofit: Retrofit = Retrofit.Builder()
            .baseUrl(BASE_URL)
            .addConverterFactory(GsonConverterFactory.create(Gson()))
            .client(client)
            .build()
        host = retrofit.create(Host::class.java)

        btn1.setOnClickListener {
            val map = HashMap<String, Hero>()
            map.put("abc", Hero.RockMan("aaaaa"))
            map.put("def", Hero.AI("ggggg", 6))
            host!!.test("kent", map).enqueue(object : Callback<User> {
                override fun onResponse(call: Call<User>, response: Response<User>) {
                    Log.d("lala", "onResponse")
                }

                override fun onFailure(call: Call<User>, t: Throwable) {
                    Log.d("lala", "onFailure t=$t")
                }
            })
        }

        btn_backup.setOnClickListener {
            val toSaveEvent = RestoreEvent.CameraEvent(1)
            val liveInfoBackupHelper = LiveInfoBackupHelper(LiveInfoBackupRepositoryImpl())
            liveInfoBackupHelper.streamID = 111
            liveInfoBackupHelper.backup(toSaveEvent)
        }

        val manager = LiveInfoRestoreAbleManager(RejoinConfig(""))

        btn_restore.setOnClickListener {
            manager.initFeature2(LiveInfoRestoreContract.Feature.Camera, object : RestoreAble<RestoreEvent.CameraEvent> {
                override fun initAction(lastModel: RestoreEvent.CameraEvent?) {
                    println("lala initAction lastModel = ${lastModel}")
                }

                override fun restoreAction(lastModel: RestoreEvent.CameraEvent?) {
                    println("lala restoreAction lastModel = ${lastModel}")
                }
            })
        }



        manager.initAction("aaa", RestoreEvent.CameraStatusEvent::class.java, object : RestoreAble<RestoreEvent.CameraStatusEvent> {
            override fun restoreAction(lastModel: RestoreEvent.CameraStatusEvent?) {
                initAction(lastModel)
                println("flag2")

                // to restore feature
            }

            override fun initAction(lastModel: RestoreEvent.CameraStatusEvent?) {
                // init
            }
        })




        println("flag3")

        val liveInfoBackupHelper = LiveInfoBackupHelper(LiveInfoBackupRepositoryImpl())
        val event = RestoreEvent.CameraStatusEvent(
            CameraMode(true, true)
        )
        liveInfoBackupHelper.backup(event)

//        println("Features.CameraStatus.key = ${RejoinContract.Features.CameraStatus.key}")
//        println("Features.CameraStatus.name = ${RejoinContract.Features.CameraStatus.name}")


    }

    interface Host {
        @POST("/test")
        fun test(
            @Query("categoryID") key: String,
            @Body map: Map<String, @JvmSuppressWildcards Hero>
        ): Call<User>

        @POST("/backup")
        fun backup(
            @Body map: Map<String, @JvmSuppressWildcards Any>
        ): Call<User>

        @POST("/{streamID}/setting")
        fun setting(
            @Path("streamID") streamID: String,
            @Body map: Map<String, @JvmSuppressWildcards Any>
        ): Call<User>

        @POST("/{streamID}/mode")
        fun mode(
            @Path("streamID") streamID: String,
            @Body map: Map<String, @JvmSuppressWildcards Any>
        ): Call<User>
    }

    private val BASE_URL = "https://api.github.com/"

    data class User(val name: String)


    companion object {
        fun launch(activity: Activity) {
            val intent = Intent(activity, RetrofitActivity::class.java)
            activity.startActivity(intent)
        }
    }

    sealed class Hero : Parcelable {
        @Parcelize
        class RockMan(val name: String) : Hero()

        @Parcelize
        class AI(val type: String, val value: Int) : Hero()
    }

    fun getTrustAllBuilder(): OkHttpClient.Builder {
        //信任所有憑證
        return try {
            val trustAllCerts = arrayOf<TrustManager>(
                object : X509TrustManager {
                    @Throws(CertificateException::class)
                    override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) {
                    }

                    @Throws(CertificateException::class)
                    override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) {
                    }

                    override fun getAcceptedIssuers(): Array<X509Certificate> {
                        return arrayOf()
                    }
                }
            )

            // Install the all-trusting trust manager
            val sslContext = SSLContext.getInstance("SSL")
            sslContext.init(null, trustAllCerts, SecureRandom())
            // Create an ssl socket factory with our all-trusting manager
            val sslSocketFactory = sslContext.socketFactory
            val builder = OkHttpClient.Builder()
            builder.sslSocketFactory(sslSocketFactory, object : X509TrustManager {
                @Throws(CertificateException::class)
                override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) {
                }

                @Throws(CertificateException::class)
                override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) {
                }

                override fun getAcceptedIssuers(): Array<X509Certificate?> {
                    return arrayOfNulls(0)
                }
            })
            builder.hostnameVerifier { hostname, session -> true }
            builder
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }
}